File: browser/src/geometry.ts
// 锚点外延 / 格尺寸 / scale — 对应 Python core/geometry.py:compute_geometry。
import type { SpriteSet, Cell, XY } from './model';

export const MARGIN = 8;

export interface Geometry {
  cellW: number;
  cellH: number;
  anchor: XY;
  scale: number;
}

/** 按"所有出现帧的最大外延"算统一格尺寸 + 锚点 + scale。
 *  外延: 上=axis_y, 下=h-axis_y, 左=axis_x, 右=w-axis_x; 取各方向最大。
 *  Python int() 截断 → Math.trunc (正数=floor)。 */
export function computeGeometry(
  frames: SpriteSet,
  cells: Iterable<Cell>,
  maxcell = 300,
  margin = MARGIN,
  scaleOverride?: number,
): Geometry {
  let mU = 1, mD = 1, mL = 1, mR = 1;
  for (const [g, i] of cells) {
    const fr = frames.get(g, i);
    if (!fr) continue;
    const [w, h] = fr.size;
    const [ax, ay] = fr.axis;
    mU = Math.max(mU, ay);
    mD = Math.max(mD, h - ay);
    mL = Math.max(mL, ax);
    mR = Math.max(mR, w - ax);
  }
  let cw = mL + mR + 2 * margin;
  let ch = mU + mD + 2 * margin;
  let anchor: XY = [mL + margin, mU + margin];
  let scale: number;
  if (scaleOverride !== undefined) {
    scale = scaleOverride; // 分组视图: 沿用整动作全局缩放
  } else {
    scale = 1.0;
    if (Math.max(cw, ch) > maxcell) scale = maxcell / Math.max(cw, ch);
  }
  cw = Math.trunc(cw * scale);
  ch = Math.trunc(ch * scale);
  anchor = [Math.trunc(anchor[0] * scale), Math.trunc(anchor[1] * scale)];
  return { cellW: cw, cellH: ch, anchor, scale };
}

// 注: 旧的 computeContentGeometry (内容居中贴合几何) 已弃用删除 —— 它丢了游戏注册锚点, 对跳跃/倒地等
// 动作帧会各自居中乱飘 (导出乱)。导出/预览改用 workflow.segmentUnionGeo (按内容 bbox 的轴相对并集,
// 见 workflow.ts): 既紧贴本组姿势 (角色大), 又按 axis 摆放保帧间连贯。

File: core/geometry.py
"""core/geometry.py — 锚点外延 / 格尺寸 / scale 计算 (单一权威, ARCHITECTURE.md §1)

把"取所有出现帧的最大外延 → 统一格尺寸 + 锚点 + scale"这套几何收成一处,
给 web (server.open_char) 和 cli (export/import) 共用。锚点口径: 每帧的"轴"
(脚底/定位点) 钉在格子固定锚点 → 所有帧"站"在同一处, AI 套形象时框架一致。
"""
from __future__ import annotations
from dataclasses import dataclass
from typing import Iterable, Optional

from core.model import SpriteSet

MARGIN = 8   # 格内留边 (上下左右各 MG)


@dataclass
class Geometry:
    """一个角色统一的格几何。

    cell_w/cell_h: 格尺寸 (已含 scale)。
    anchor:        (x, y) 轴在格内的钉点 (已含 scale)。
    scale:         为不超 maxcell 而做的缩放 (1.0 = 不缩)。
    """
    cell_w: int
    cell_h: int
    anchor: tuple[int, int]
    scale: float


def compute_geometry(frames: SpriteSet, cells: Iterable[tuple[int, int]],
                     maxcell: int = 300, margin: int = MARGIN,
                     scale_override: Optional[float] = None) -> Geometry:
    """按"所有出现帧的最大外延"算统一格尺寸 + 锚点 + scale。

    frames: SpriteSet | dict[(g,i)] = Frame (查帧)。
    cells:  可迭代 (group, image) — 参与几何的帧坐标 (通常是若干动作的去重并集)。
    maxcell: 格长边上限, 超出按比例缩 (默认 300, server 口径; cli 用 320)。

    外延: 上=axis_y, 下=h-axis_y, 左=axis_x, 右=w-axis_x; 取各方向最大值。
    返回 Geometry。
    """
    mU = mD = mL = mR = 1
    for (g, i) in cells:
        fr = frames.get(g, i)        # frames: SpriteSet (get(group, image))
        if not fr:
            continue
        w, h = fr.size
        ax, ay = fr.axis
        mU = max(mU, ay); mD = max(mD, h - ay)
        mL = max(mL, ax); mR = max(mR, w - ax)
    cw = mL + mR + 2 * margin
    ch = mU + mD + 2 * margin
    anchor = (mL + margin, mU + margin)
    if scale_override is not None:   # 分组视图: 沿用整动作全局缩放(角色同比例可拼回), 格子收到本组外延
        scale = scale_override
    else:
        scale = 1.0
        if max(cw, ch) > maxcell:
            scale = maxcell / max(cw, ch)
    cw = int(cw * scale)
    ch = int(ch * scale)
    anchor = (int(anchor[0] * scale), int(anchor[1] * scale))
    return Geometry(cell_w=cw, cell_h=ch, anchor=anchor, scale=scale)
